doc: security: update docs for Mbed TLS 4.1.0 (#29023)
nrfconnect/sdk-nrf: doc/nrf/libraries/security/nrf_security.rst

Reviewer: Would've been useful to move a file (e.g. `nrf_security/index.rst`) instead in order to have a smaller diff

############

The nRF Security subsystem (``nrf_security``) integrates cryptographic services for SoCs from Nordic Semiconductor.

Overview
********

The nRF Security subsystem provides:

* :ref:`PSA Crypto APIs <ug_psa_certified_api_overview_crypto>` through Oberon PSA Crypto (TF-PSA-Crypto)
* TLS, DTLS, and X.509 support through `Mbed TLS`_
* Hardware acceleration through dedicated cryptographic libraries on selected SoCs (``nrf_cc3xx``, CRACEN), with binary versions of the libraries listed in :ref:`nrfxlib:crypto`
* Software fallbacks when hardware acceleration is unavailable (``nrf_oberon``)
* A PSA driver abstraction layer enabling simultaneous use of hardware and software implementations
* Compatibility with the specific Mbed TLS version included in the |NCS| through `sdk-mbedtls`_
* Integration logic for the Oberon PSA Crypto core (`sdk-oberon-psa-crypto`_)
* Source code for the CRACEN driver
* Integration with the |NCS| build system

The nRF Security subsystem can interface with the :ref:`nrf_cc3xx_mbedcrypto_readme`.
This library conforms to the specific revision of Mbed TLS that is supplied through the |NCS|.

.. _nrf_security_config:

Configuration
*************

> You can enable nRF Security by setting the :kconfig:option:`CONFIG_NRF_SECURITY` Kconfig option.

Reviewer 2: ```suggestion
To enable nRF Security, set the :kconfig:option:`CONFIG_NRF_SECURITY` Kconfig option.
```

Overview
********

The nRF Security subsystem provides:

* The PSA Crypto API through :ref:`Oberon PSA Crypto<ug_psa_certified_api_overview_crypto>`.
* TLS, DTLS, and X.509 support through `Mbed TLS`_.
* Hardware acceleration through dedicated cryptographic libraries on selected SoCs (``nrf_cc3xx``, CRACEN), with binary versions of the libraries listed in :ref:`nrfxlib:crypto`.
* Software fallbacks when hardware acceleration is unavailable (``nrf_oberon``).
* A PSA driver abstraction layer that enables simultaneous use of hardware and software implementations.
* Compatibility with the specific Mbed TLS version included in the |NCS| through `sdk-mbedtls`_.
* Integration logic for the Oberon PSA Crypto core (`sdk-oberon-psa-crypto`_).
* Source code for the CRACEN driver.
* Integration with the |NCS| build system.

The nRF Security subsystem can interface with the :ref:`nrf_cc3xx_mbedcrypto_readme`.
This library conforms to the specific revision of Mbed TLS that is supplied through the |NCS|.

.. _nrf_security_config:

Configuration
*************

To enable nRF Security, enable the :kconfig:option:`CONFIG_PSA_CRYPTO` Kconfig option.

> On Nordic Semiconductor's Arm cores, this automatically enables :kconfig:option:`CONFIG_NRF_SECURITY`, which builds and configures the nRF Security subsystem.

Reviewer: `builds and configures the nRF Security subsystem` doesn't sound quite right (because nRF Security is not your regular subsystem). `which automatically enables nRF Security` would be better, but then it would be a repetition.
```suggestion
On Nordic Semiconductor's Arm cores, this automatically enables :kconfig:option:`CONFIG_NRF_SECURITY`.
```

PSA Crypto
==========

Cryptographic operations are enabled with the :kconfig:option:`CONFIG_PSA_CRYPTO` Kconfig option.

.. ncs-include:: ../../security/crypto/driver_config.rst
   :start-after: psa_crypto_support_def_start
   :end-before: psa_crypto_support_def_end

For more information, see :ref:`psa_crypto_support`.
For the list of supported crypto features, see :ref:`ug_crypto_supported_features`.

Depending on the implementation you are using, the |NCS| builds nRF Security using different versions of the PSA Crypto API.

.. ncs-include:: ../../security/psa_certified_api_overview.rst
   :start-after: psa_crypto_support_tfm_build_start
   :end-before: psa_crypto_support_tfm_build_end

.. _nrf_security_tls_x509_config:

Mbed TLS (TLS and X.509)
========================

Enable Mbed TLS only when your application needs TLS, DTLS, or X.509 certificate handling.
To enable TLS and X.509 support, set the :kconfig:option:`CONFIG_MBEDTLS` Kconfig option.

.. note::

>    For all other cryptographic operations, use the `PSA Crypto`_ configuration through :kconfig:option:`CONFIG_PSA_CRYPTO`.

Reviewer: Not sure what you wanted to say?
```suggestion
   For all other cryptographic operations, use the `PSA Crypto`_ API by enabling :kconfig:option:`CONFIG_PSA_CRYPTO`.
```

Author: Wanted to point to the previous section.

>    The subsystem uses the Mbed TLS integration from Zephyr, but replaces TF-PSA-Crypto used in Zephyr with Oberon PSA Crypto.

Reviewer: ```suggestion
   nRF Security uses the Mbed TLS integration from Zephyr, but replaces the TF-PSA-Crypto repository used in Zephyr with Oberon PSA Crypto.
```

   Some TF-PSA-Crypto Kconfig options still use the ``MBEDTLS_`` prefix even when :kconfig:option:`CONFIG_MBEDTLS` is disabled.
   These options configure the crypto stack, not Mbed TLS.

When using TLS, enable :kconfig:option:`CONFIG_MBEDTLS_ENABLE_HEAP` and adjust :kconfig:option:`CONFIG_MBEDTLS_HEAP_SIZE` for your workload.

For the Mbed TLS version included in the |NCS|, see :ref:`security_index`.

Dependencies
************

The nRF Security subsystem uses the following |NCS| modules:

> * `sdk-oberon-psa-crypto`_ - Oberon PSA Crypto (PSA Crypto Core and TF-PSA-Crypto implementation; replaces Zephyr's TF-PSA-Crypto in the |NCS|)

Reviewer: If we want to be more precise:
```suggestion
* `sdk-oberon-psa-crypto`_ - Oberon PSA Crypto (PSA Crypto Core and TF-PSA-Crypto implementation; replaces Zephyr's TF-PSA-Crypto when nRF Security is enabled)
```
(because there are legit cases when using NCS where nRF Security won't be enabled, on non-Arm cores for instance)
\+
```suggestion
* `sdk-oberon-psa-crypto`_ - Oberon PSA Crypto (PSA Crypto core and TF-PSA-Crypto implementation; replaces Zephyr's TF-PSA-Crypto in the |NCS|)
```

Author: Core is capitalized in https://github.com/nrfconnect/sdk-oberon-psa-crypto, so I stick to that

* `sdk-nrfxlib`_ - Prebuilt cryptographic libraries (CryptoCell, Oberon software libraries)
* `sdk-mbedtls`_ - TLS, DTLS, and X.509 support (only when :kconfig:option:`CONFIG_MBEDTLS` is enabled)

> * `sdk-trusted-firmware-m`_ - Security by separation with ARM TrustZone and TF-M Crypto Service (only when :kconfig:option:`CONFIG_BUILD_WITH_TFM` is enabled)

Reviewer: ```suggestion
* `sdk-trusted-firmware-m`_ - Security by separation with ARM TrustZone and TF-M Crypto service (only when TF-M is enabled by building for ``*/ns`` board targets)
```

.. _nrf_security_readme:
.. _nrf_security:

nRF Security
############

The nRF Security subsystem (``nrf_security``) integrates cryptographic services for SoCs from Nordic Semiconductor.

Overview
********

The nRF Security subsystem provides:

> * PSA Crypto APIs through :ref:`Oberon PSA Crypto<ug_psa_certified_api_overview_crypto>`

Reviewer: ```suggestion
* The PSA Crypto API through :ref:`Oberon PSA Crypto<ug_psa_certified_api_overview_crypto>`
```

* TLS, DTLS, and X.509 support through `Mbed TLS`_
* Hardware acceleration through dedicated cryptographic libraries on selected SoCs (``nrf_cc3xx``, CRACEN), with binary versions of the libraries listed in :ref:`nrfxlib:crypto`
* Software fallbacks when hardware acceleration is unavailable (``nrf_oberon``)
* A PSA driver abstraction layer enabling simultaneous use of hardware and software implementations
* Compatibility with the specific Mbed TLS version included in the |NCS| through `sdk-mbedtls`_
* Integration logic for the Oberon PSA Crypto core (`sdk-oberon-psa-crypto`_)
* Source code for the CRACEN driver
* Integration with the |NCS| build system

The nRF Security subsystem can interface with the :ref:`nrf_cc3xx_mbedcrypto_readme`.
This library conforms to the specific revision of Mbed TLS that is supplied through the |NCS|.

.. _nrf_security_config:

Configuration
*************

> To enable nRF Security, set the :kconfig:option:`CONFIG_NRF_SECURITY` Kconfig option.

Reviewer: This is wrong, `CONFIG_NRF_SECURITY` is now promptless.
```suggestion
To enable nRF Security, enable the :kconfig:option:`CONFIG_PSA_CRYPTO` Kconfig option.
:kconfig:option:`CONFIG_NRF_SECURITY` will be automatically enabled as a result of that on Nordic Semiconductor's Arm-based board targets.
```

.. note::
   |zephyr_tf_psa_crypto_incompatibility_note|

>    Read :ref:`ug_crypto_architecture_implementation_standards` for more information about implementations in the |NCS| that support Nordic Semiconductor's Arm SoCs.

Reviewer: ```suggestion
   Read :ref:`ug_crypto_architecture_implementation_standards` for more information about implementations in the |NCS| that support Nordic Semiconductor's Arm cores.
```
Cores rather than SoCs because for example the 54L15 is a SoC and it has multiple cores, `cpuapp` being an Arm one and `cpuflpr` a RISC-V one.

PSA Crypto
==========

To use cryptographic operations, enable the :kconfig:option:`CONFIG_PSA_CRYPTO` Kconfig option.

> On Nordic Semiconductor's Arm SoCs, this automatically enables :kconfig:option:`CONFIG_NRF_SECURITY`, which builds and configures the nRF Security subsystem.

Reviewer: Cores again, not SoCs, I guess there will be more places to update

> On Nordic Semiconductor's Arm SoCs, this automatically enables :kconfig:option:`CONFIG_NRF_SECURITY`, which builds and configures the nRF Security subsystem.

Reviewer: So here it's correct, we'll need to somehow combine it with your sentence that says how to enable nRF Security.

.. ncs-include:: ../../security/crypto/driver_config.rst
   :start-after: psa_crypto_support_def_start
   :end-before: psa_crypto_support_def_end

> The PSA Crypto API is enabled by default when you enable nRF Security.

Reviewer: Can remove this as you aren't really meant to enable nRF Security directly yourself.

For more information, see :ref:`psa_crypto_support`.
For the list of supported crypto features, see :ref:`ug_crypto_supported_features`.

Depending on the implementation you are using, the |NCS| builds nRF Security using different versions of the PSA Crypto API.

.. ncs-include:: ../../security/psa_certified_api_overview.rst
   :start-after: psa_crypto_support_tfm_build_start
   :end-before: psa_crypto_support_tfm_build_end

.. _nrf_security_tls_x509_config:

TLS and X.509
=============

Enable Mbed TLS only when your application needs TLS, DTLS, or X.509 certificate handling.
For all other cryptographic operations, use the `PSA Crypto`_ configuration through :kconfig:option:`CONFIG_PSA_CRYPTO`.

To enable TLS and X.509 support, set the :kconfig:option:`CONFIG_MBEDTLS` Kconfig option.

> Enabling :kconfig:option:`CONFIG_MBEDTLS_TLS_LIBRARY` selects :kconfig:option:`CONFIG_MBEDTLS` and the X.509 library (:kconfig:option:`CONFIG_MBEDTLS_X509_LIBRARY`).

Reviewer: I want to get rid of `CONFIG_MBEDTLS_TLS_LIBRARY` and `CONFIG_MBEDTLS_X509_LIBRARY` which are leftovers at this point so let's not mention them. `CONFIG_MBEDTLS` is what users need to know about and enable.

.. note::
   Some TF-PSA-Crypto Kconfig options still use the ``MBEDTLS_`` prefix even when :kconfig:option:`CONFIG_MBEDTLS` is disabled.

>    Enabling these options configures the crypto stack, not the TLS library.

Reviewer: ```suggestion
   These options configure the crypto stack, not Mbed TLS.
```

> When using TLS, enable :kconfig:option:`CONFIG_MBEDTLS_ENABLE_HEAP` and adjust :kconfig:option:`CONFIG_MBEDTLS_HEAP_SIZE` for your workload.

Reviewer: Hmm where did you take this from? I'm not certain about the validity of this advice

Author: CONFIG_MBEDTLS_HEAP_SIZE

[memory_optimization.rst](https://github.com/nrfconnect/sdk-nrf/blob/main/doc/nrf/protocols/matter/getting_started/memory_optimization.rst?plain=1#L425) 

[aws_fota.rst](https://github.com/nrfconnect/sdk-nrf/blob/main/doc/nrf/libraries/networking/aws_fota.rst?plain=1#L196)

CONFIG_MBEDTLS_ENABLE_HEAP

[CMakeLists.txt of nrf_security](https://github.com/nrfconnect/sdk-nrf/blob/main/subsys/nrf_security/src/zephyr/CMakeLists.txt#L23)

Reviewer: Okay, I think it's a bit misguided because these two Kconfig options configure the heap for both TF-PSA-Crypto and Mbed TLS now, so they're not specific to TLS.
The advice to adjust `CONFIG_MBEDTLS_HEAP_SIZE` when `CONFIG_MBEDTLS_ENABLE_HEAP` is enabled is still valid, but could probably rather be placed in the `PSA Crypto` section (or in a different, common section).

For the Mbed TLS version included in the |NCS|, see :ref:`security_index`.

Dependencies
************

The nRF Security subsystem uses the following |NCS| modules:

> * `sdk-oberon-psa-crypto`_ - Oberon PSA Crypto (TF-PSA-Crypto) PSA Crypto core

Reviewer: ```suggestion
* `sdk-oberon-psa-crypto`_ - Oberon PSA Crypto (PSA Crypto core and TF-PSA-Crypto implementation)
```

* `sdk-nrfxlib`_ - Prebuilt cryptographic libraries (CryptoCell, Oberon software libraries)
* `sdk-mbedtls`_ - TLS, DTLS, and X.509 support (only when :kconfig:option:`CONFIG_MBEDTLS` is enabled)

> * `sdk-trusted-firmware-m`_ - TF-M Crypto Service (only when :kconfig:option:`CONFIG_BUILD_WITH_TFM` is enabled)

Reviewer: well TF-M is more than just the crypto service

> The subsystem selects :ref:`cryptographic drivers <crypto_drivers>` based on Kconfig options such as :kconfig:option:`CONFIG_PSA_CRYPTO_DRIVER_CC3XX`, :kconfig:option:`CONFIG_PSA_CRYPTO_DRIVER_CRACEN`, and :kconfig:option:`CONFIG_PSA_CRYPTO_DRIVER_OBERON`.

Reviewer: ```suggestion
The subsystem uses :ref:`cryptographic drivers <crypto_drivers>` based on Kconfig options such as :kconfig:option:`CONFIG_PSA_CRYPTO_DRIVER_CC3XX`, :kconfig:option:`CONFIG_PSA_CRYPTO_DRIVER_CRACEN`, and :kconfig:option:`CONFIG_PSA_CRYPTO_DRIVER_OBERON`.
```
?
